### PR TITLE
DataLoader: properly diagnose exceeding file descriptor limit

### LINKED ISSUE #973: file_descriptor sharing strategy may be leaking FDs, resulting in DataLoader causing `RuntimeError: received 0 items of ancdata`
**Editorial note:** If you are having this problem, try running `torch.multiprocessing.set_sharing_strategy('file_system')` right after your import of torch

----

I am using a `DataLoader` in my code with a custom `Dataset` class, and it worked fine during training for several epochs. However, when testing my model, after a bit less than 1k iterations, I'm getting the following error:

```
RuntimeError                              Traceback (most recent call last)
/home/jfsantos/src/pytorch_models/test_model.py in <module>()
     82
     83 print('Generating samples...')
---> 84 for k, batch in tqdm(enumerate(test_loader)):
     85     f = G_test.audio_paths[k]
     86     spec, phase = spectrogram_from_file(f, window=window, step=step)

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/site-packages/tqdm/_tqdm.py in __iter__(self)
    831 """, fp_write=getattr(self.fp, 'write', sys.stderr.write))
    832
--> 833             for obj in iterable:
    834                 yield obj
    835                 # Update and print the progressbar.

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/site-packages/torch/utils/data/dataloader.py in __next__(self)
    166         while True:
    167             assert (not self.shutdown and self.batches_outstanding > 0)
--> 168             idx, batch = self.data_queue.get()
    169             self.batches_outstanding -= 1
    170             if idx != self.rcvd_idx:

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/multiprocessing/queues.py in get(self)
    343             res = self._reader.recv_bytes()
    344         # unserialize the data after having released the lock
--> 345         return ForkingPickler.loads(res)
    346
    347     def put(self, obj):

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/site-packages/torch/multiprocessing/reductions.py in rebuild_storage_fd(cls, df, size)
     68         fd = multiprocessing.reduction.rebuild_handle(df)
     69     else:
---> 70         fd = df.detach()
     71     try:
     72         storage = storage_from_cache(cls, fd_id(fd))

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/multiprocessing/resource_sharer.py in detach(self)
     56             '''Get the fd.  This should only be called once.'''
     57             with _resource_sharer.get_connection(self._id) as conn:
---> 58                 return reduction.recv_handle(conn)
     59
     60

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/multiprocessing/reduction.py in recv_handle(conn)
    179         '''Receive a handle over a local connection.'''
    180         with socket.fromfd(conn.fileno(), socket.AF_UNIX, socket.SOCK_STREAM) as s:
--> 181             return recvfds(s, 1)[0]
    182
    183     def DupFd(fd):

/home/jfsantos/anaconda3/envs/pytorch/lib/python3.5/multiprocessing/reduction.py in recvfds(sock, size)
    158             if len(ancdata) != 1:
    159                 raise RuntimeError('received %d items of ancdata' %
--> 160                                    len(ancdata))
    161             cmsg_level, cmsg_type, cmsg_data = ancdata[0]
    162             if (cmsg_level == socket.SOL_SOCKET and

RuntimeError: received 0 items of ancdata
```

However, if I just do `idxs = [k for k, batch in tqdm(enumerate(test_loader))]` I do **not** have this issue.

I do not have any idea on how to test it as my knowledge of how PyTorch does this is currently very limited, but I could help debug this given some instructions. Does anyone have any idea on where I could start?

### PR DESCRIPTION
**Pull request by Baranowski**
Fixes #973

Common failure scenario:
* DataLoader creates workers and communicates with them through SHMs
* Workers send back through an AF_UNIX socket file descriptors to SHMs containing data
* The limit of open files gets fully used
* A FD gets stripped from a socket message coming back from a worker, without the worker knowing this.
* This causes a `RuntimeError: received 0 items of ancdata` in the standard `multiprocessing` package
* The exception is not handled by PyTorch and so is presented to the users.

After this change the user will see

```
Traceback (most recent call last):
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 761, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
  File "/home/wbaranowski/git/Quansight/pytorch/torch/multiprocessing/reductions.py", line 294, in rebuild_storage_fd
    fd = df.detach()
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/multiprocessing/resource_sharer.py", line 58, in detach
    return reduction.recv_handle(conn)
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/multiprocessing/reduction.py", line 184, in recv_handle
    return recvfds(s, 1)[0]
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/multiprocessing/reduction.py", line 162, in recvfds
    len(ancdata))
RuntimeError: received 0 items of ancdata

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 787, in _try_get_data
    fs = [tempfile.NamedTemporaryFile() for i in range(10)]
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 787, in <listcomp>
    fs = [tempfile.NamedTemporaryFile() for i in range(10)]
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/tempfile.py", line 551, in NamedTemporaryFile
    (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/home/wbaranowski/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/tempfile.py", line 262, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
OSError: [Errno 24] Too many open files: '/tmp/tmpnx_f6v_f'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_shm_leak.py", line 56, in <module>
    worker_init_fn=worker_init_fn
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 345, in __next__
    data = self._next_data()
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 861, in _next_data
    idx, data = self._get_data()
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 828, in _get_data
    success, data = self._try_get_data()
  File "/home/wbaranowski/git/Quansight/pytorch/torch/utils/data/dataloader.py", line 791, in _try_get_data
    "Too many open files. Communication with the"
RuntimeError: Too many open files. Communication with the workers is no longer possible. Please increase the limit using `ulimit -n` in the shell or change the sharing strategy by calling `torch.multiprocessing.set_sharing_strategy('file_system')` at the beginning of your code
```

### CODE REVIEW

**Comment by Baranowski** [comment]
TL;DR:

- The key issue is most likely in the Python packages `multiprocessing`, `socket` or even the C implementation of an AF_UNIX socket. IMO it's best to work around it in PyTorch
- Adding a test would be challenging because the size of the data used in a test would fully depend on the host's limits, or alternatively it might mess up ulimit for subsequent tests if the exit is not clean.


I managed to reproduce the same behaviour outside PyTorch with the following script:

```python
#!/usr/bin/env python3
import sys
import socket
import os
import array
import shutil
import socket


if len(sys.argv) != 4:
    print("Usage: ", sys.argv[0], " tmp_dirname iteration (send|recv)")
    sys.exit(1)

if __name__ == '__main__':
    dirname = sys.argv[1]
    sock_path = dirname + "/sock"
    iterations = int(sys.argv[2])
    def dummy_path(i):
        return dirname + "/" + str(i) + ".dummy"


    if sys.argv[3] == 'send':
        while not os.path.exists(sock_path):
            pass
        client = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
        client.connect(sock_path)
        for i in range(iterations):
            fd = os.open(dummy_path(i), os.O_WRONLY | os.O_CREAT)
            ancdata = array.array('i', [fd])
            msg = bytes([i % 256])
            print("Sending fd ", fd, " (iteration #", i, ")")
            client.sendmsg([msg], [(socket.SOL_SOCKET, socket.SCM_RIGHTS, ancdata)])


    else:
        assert sys.argv[3] == 'recv'

        if os.path.exists(dirname):
            raise Exception("Directory exists")

        os.mkdir(dirname)

        print("Opening socket...")
        server = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
        server.bind(sock_path)

        print("Listening...")
        for i in range(iterations):
            a = array.array('i')
            msg, ancdata, flags, addr = server.recvmsg(1, socket.CMSG_SPACE(a.itemsize))
            assert(len(ancdata) == 1)
            cmsg_level, cmsg_type, cmsg_data = ancdata[0]
            a.frombytes(cmsg_data)
            print("Received fd ", a[0], " (iteration #", i, ")")

        shutil.rmtree(dirname)
```

Steps to reproduce:

1. Run two shells and send lower file descriptor limit in the receiving one:
```
(shell1) ulimit -n 1020
(shell2) ulimit -n 1022
```

2. Run the script above with the `recv` option in the first shell
```
./test_socket.py sock_tmp 1017 recv
```

3. Run the script with the `send` option in the second shell:
```
./test_socket.py sock_tmp 1017 send
```

The key problem here is that when the root process is about to receive an FD over a socket from a worker, that FD would exceed the open file descriptors limit. To mitigate that, the FD is silently removed from the message and the `MSG_CTRUNC` flag is set. (The flag might be a bit misleading since the manpage says that it _indicates that some control data were discarded due to lack of space in the buffer for ancillary data_. It doesn't really matter, though, since the standard `multiprocessing` package ignores it.)

I haven't verified it but I suspect that this behaviour of `recvmsg` in Python's `socket` is consistent with `recvmsg` in C. I'm not sure it's the desired behaviour. AFAICT, there is no indication of an issue anywhere, other than the `MSG_CTRUNC` flag. Anyway, digging into the C implementation of recvmsg would be fun but outside of the scope of the PyTorch issue.

The worker itself does not exceed the limit in the common scenario. If you tweak the ulimit and number of the FDs received/sent in the script above, instead of the `RuntimeError: received 0 items of ancdata` you might get `OSError: Too many open files` on the receiver/sender or both. During my testing of PyTorch I have seen that `OSError` passed through the DataLoader a couple times but I couldn't reproduce it afterwards.


I would love to add a test for this change, but I'm worried that if I lower ulimit for a single test, it might affect subsequent tests if my test exits in an unexpected way. An alternative approach would be to send as many tensors as ulimit allows. But since that environment setting can have very wide range, it might cause some unexpected issues.

**Comment by ezyang** [comment]
@SsnL do you think you would be able to review this PR? (Baranowski bug me again in a day if there's no response)

**Comment by ezyang** [comment]
Re testing, you should just run these tests in a completely separate process from the rest of the tests. There are some examples of how to do this, see for example `TestCuda.test_cuda_init_race` (in general, grep for subprocess in the test suite)

File: fd_leak_comment.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+```
+
+3. Run the script with the `send` option in the second shell:
+```
+./test_socket.py sock_tmp 1017 send
+```
+
+I haven't verified it but I suspect that this behaviour of `recvmsg` in Python's `multiprocessing` is consistent with `recvmsg` in C. I'm not sure it's the desired behaviour. AFAICT, there is no indication of an issue anywhere, other than the `MSG_CTRUNC` flag.
+
+Anyway
```
Comment (ezyang) [inline fd_leak_comment.md]: ??

File: fd_leak_comment.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+Nothing actionable in this comment. Just a findings update for the curious and to help me gather my thoughts and remember the key facts.
```
Comment (ezyang) [inline fd_leak_comment.md]: This probably shouldn't live as a top level md in the repo. Maybe put it as a comment near the relevant code.

**Comment by ezyang** [comment]
Code looks generally reasonable, but I'd suggest spending a little time trying to check in the very nice test you've written.

File: fd_leak_comment.md (hunk 2 above)
Comment (Baranowski) [reply fd_leak_comment.md]: Sorry, I added this file to the PR by mistake. It was supposed to stay only in my local working branch.

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -774,6 +774,27 @@ def _try_get_data(self, timeout=_utils.MP_STATUS_CHECK_INTERVAL):
...
+            import errno
+            try:
+                # If we are here, it might be because the FDs we were about to
+                # receive over a socket would exceed the open FDs limit. Let's
+                # test it the obvious way.
+
+                # Raise an exception if we are this close to the FDs limit.
+                # Apparently, trying to open only one file is not a sufficient
+                # test but I don't understand why.
+                fds_limit_margin = 10
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: The reason you need more than 1 could be that you need multiple FDs to receive that message from the socket. I imagine that it'd be impossible to get the correct # of FD needed. Maybe 10 is okay.

**Review by ssnl — COMMENTED** [review]
Nice analysis! 

I think the note you wrote deserves a place in the code base. You could add it closer with the actual relevant code, or just as inline comments. You can grep for `NOTE` in the codebase for examples. Then you can cross-ref it in `dataloader.py:779` comment block.

On test: 
You might find it hard to convert your script into a proper pytorch test, especially when it does `ulimit -n`. It would be okay, but please keep the nice script in the note. :)

File: test/test_dataloader.py [hunk 4]
```diff
@@ -777,6 +777,43 @@ def test_error_in_init(self):
         with self.assertRaisesRegex(RuntimeError, 'Error in worker_init_fn'):
             list(iter(loader))
 
+    @unittest.skipIf(IS_WINDOWS, "Only supported or needed on Unix")
+    def test_fd_limit_exceeded(self):
```
Comment (SsnL) [inline test/test_dataloader.py]: could you add a ref to the note? :)

**Comment by SsnL** [comment]
macos failure seems real.

**Comment by SsnL** [comment]
`NameError: name 'IS_MACOS' is not defined`

**Review by ssnl — APPROVED** [review]
Thanks!

**Comment by Baranowski** [comment]
Thanks. Sorry for the unnecessary back-and-forth at the end.

**Comment by SsnL** [comment]
No, thank you for the analysis and fix!

**Comment by ezyang** [comment]
Internal CI failure

```
  warnings.warn(err_msg)
test_fd_limit_exceeded (test_dataloader.TestDataLoader) ... I0319 21:23:47.775116 749361 init.cc:561] Skip logging in unit test environment for event: tensor.create
Traceback (most recent call last):
  File "<string>", line 3, in <module>
ModuleNotFoundError: No module named 'torch.utils'
ERROR

======================================================================
ERROR: test_fd_limit_exceeded (test_dataloader.TestDataLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/dataloader#binary,link-tree/test_dataloader.py", line 816, in test_fd_limit_exceeded
    """])
  File "/usr/local/fbcode/platform007/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/local/fbcode/platform007/bin/python3.7', '-c', 'import torch\nimport resource\nfrom torch.utils.data import DataLoader, IterableDataset\n\nclass RandomDataset(IterableDataset):\n    def __init__(self, len, size):\n        super(RandomDataset).__init__()\n        self.len = len\n        self.size = size\n\n    def __iter__(self):\n        return self\n\n    def __next__(self):\n        if self.len <= 0:\n            raise StopIteration\n        self.len -= 1\n        return torch.randn(self.size)\n\ntry:\n    keep_fds_alive = []\n    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))\n    for random_t in DataLoader(RandomDataset(200, (2,2)),\n                               num_workers=1):\n      random_t.max(dim=0)\n      keep_fds_alive.append(random_t)\nexcept RuntimeError as e:\n    assert "ulimit -n" in str(e)\n    assert "set_sharing_strategy" in str(e)\nelse:\n    assert False, "Expected a RuntimeError with keywords ulimit and set_sharing_strategy"\n']' returned non-zero exit status 1.

----------------------------------------------------------------------
R
```

Maybe it's just something easy?

**Comment by Baranowski** [comment]
I guess AF_UNIX sockets were not used in this test for some reason.

Now that I think of it, this test should not enforce that DataLoader fails. Rather, it should enforce that DataLoader **either** works correctly or fails in the known way. I will fix that but I'm not sure I will manage to do it today.

**Comment by ezyang** [comment]
Thanks, no hurry.

**Comment by ezyang** [comment]
It still seems to be failing in the same way

**Comment by Baranowski** [comment]
Darn. I'm afraid that without the stderr of the subprocess I won't be able to debug it.

**Comment by ezyang** [comment]
Full test output:

```
CalledProcessError: Command '['/usr/local/fbcode/platform007/bin/python3.7', '-c', 'import torch\nimport resource\nfrom torch.utils.data import DataLoader, IterableDataset\n\nclass RandomDataset(IterableDataset):\n    def __init__(self, len, size):\n        super(RandomDataset).__init__()\n        self.len = len\n        self.size = size\n\n    def __iter__(self):\n        return self\n\n    def __next__(self):\n        if self.len <= 0:\n            raise StopIteration\n        self.len -= 1\n        return torch.randn(self.size)\n\ntry:\n    keep_fds_alive = []\n    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))\n    for random_t in DataLoader(RandomDataset(200, (2,2)),\n                               num_workers=1):\n      random_t.max(dim=0)\n      keep_fds_alive.append(random_t)\nexcept RuntimeError as e:\n    assert "ulimit -n" in str(e)\n    assert "set_sharing_strategy" in str(e)\n']' returned non-zero exit status 1.
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/dataloader#binary,link-tree/test_dataloader.py", line 814, in test_fd_limit_exceeded
    """])
  File "/usr/local/fbcode/platform007/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)

WARNING: Logging before InitGoogleLogging() is written to STDERR
I0324 10:11:37.182368 979065 init.cc:561] Skip logging in unit test environment for event: torch.python.import
I0324 10:11:37.720602 979065 init.cc:561] Skip logging in unit test environment for event: c10d.python.import
I0324 10:11:37.835006 979065 init.cc:561] Skip logging in unit test environment for event: torch.script.compile
/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/dataloader#binary,link-tree/torch/_utils_internal.py:79: DeprecationWarning: This is a NOOP in python >= 3.7, its just too dangerous with how we write code at facebook. Instead we patch os.fork and multiprocessing which can raise exceptions if a deadlock would happen. 
  threadSafeForkRegisterAtFork()
/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/dataloader#binary,link-tree/test_dataloader.py:32: UserWarning: psutil not found. Some critical data loader tests relying on it (e.g., TestDataLoader.test_proper_exit) will not run.
  warnings.warn(err_msg)
test_fd_limit_exceeded (test_dataloader.TestDataLoader) ... I0324 10:11:38.033820 979065 init.cc:561] Skip logging in unit test environment for event: tensor.create
Traceback (most recent call last):
  File "<string>", line 3, in <module>
ModuleNotFoundError: No module named 'torch.utils'
ERROR

======================================================================
ERROR: test_fd_limit_exceeded (test_dataloader.TestDataLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/caffe2/test/dataloader#binary,link-tree/test_dataloader.py", line 814, in test_fd_limit_exceeded
    """])
  File "/usr/local/fbcode/platform007/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/local/fbcode/platform007/bin/python3.7', '-c', 'import torch\nimport resource\nfrom torch.utils.data import DataLoader, IterableDataset\n\nclass RandomDataset(IterableDataset):\n    def __init__(self, len, size):\n        super(RandomDataset).__init__()\n        self.len = len\n        self.size = size\n\n    def __iter__(self):\n        return self\n\n    def __next__(self):\n        if self.len <= 0:\n            raise StopIteration\n        self.len -= 1\n        return torch.randn(self.size)\n\ntry:\n    keep_fds_alive = []\n    resource.setrlimit(resource.RLIMIT_NOFILE, (100, 100))\n    for random_t in DataLoader(RandomDataset(200, (2,2)),\n                               num_workers=1):\n      random_t.max(dim=0)\n      keep_fds_alive.append(random_t)\nexcept RuntimeError as e:\n    assert "ulimit -n" in str(e)\n    assert "set_sharing_strategy" in str(e)\n']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 1 test in 0.430s

FAILED (errors=1)
"/usr/bin/time:
MaxMemory 1396940
Inputs 271824
Outputs 0"
```

**Comment by Baranowski** [comment]
Sorry, previously I ignored the key information which was already present in your snippet:
```
Traceback (most recent call last):
  File "<string>", line 3, in <module>
ModuleNotFoundError: No module named 'torch.utils'
```

Unfortunately without knowing anything about the FB internal testing environment, I cannot explain why that module is available to the test process but not to the subprocess. According to Python docs, the subprocess should inherit environment of the parent.

I will add a simple workaround in a sec and let you decide whether to go with that.

**Comment by ezyang** [comment]
Oh, that diagnosis is helpful. I may have an idea...

**Comment by ezyang** [comment]
Sorry about the delay. What I think the internal problem is, is that we're invoking python subprocess and it's not picking up the same python environment as the parent process. I'm pretty sure we've done this correctly elsewhere and I need to figure out how; haven't had a chance to yet.

**Comment by rgommers** [comment]
> we're invoking python subprocess and it's not picking up the same python environment as the parent process

I just ran into this somewhere else. A couple of things that may help:
- Using `sys.executable` instead of `python` to invoke Python in the subprocess
- If it's a conda env, it needs activation
- If it's a virtualenv it's easier, but still `PATH` and `PYTHONHOME` can be different in the subprocess (see e.g. [this SO discussion](https://stackoverflow.com/questions/8052926/running-subprocess-within-different-virtualenv-with-python)) so setting those or `source activate`ing may help.

**Comment by ezyang** [comment]
We're using `sys.executable`, but the way Python tests run internally is speshul so I wonder if that's related.

**Comment by SsnL** [comment]
`subprocess.check_output` should by default pass over the env vars. Could we add `pprint(os.environ)` in the script to make sure that env vars are the problem?

**Comment by ezyang** [comment]
As best as I can tell, invoking python from subprocess has never actually worked in our internal build environment (I *thought* it might work, because there were many subprocess invocations that weren't guarded, but as best I can tell they all accidentally weren't run at all in fbcode for various other reasons). So let's just skip this test when we run it internally. You can case on this using `IS_SANDCASTLE`, just do a skipIf on your test. Sorry about the hold up.

**Comment by Baranowski** [comment]
@ezyang Done, this should be ready to merge. The CI failure looks unrelated to me.
